{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c9e02853",
   "metadata": {},
   "source": [
    "# info\n",
    "This script generates a file with filtered DEGs derived from patient-specific DEG calling."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fce7cc84",
   "metadata": {},
   "source": [
    "Filters are:\n",
    "1. Filter on P < 0.05 and Q < 0.1 for DESeq2.\n",
    "2. Filter out genes that in the comparison do not cross the 2 TPM barrier.\n",
    "3. Discrete abs log2 FC > 1\n",
    "\n",
    "Coherent DEGs are the ones that pass all three filters in n-1 patients."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f1e61540",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas, os, numpy, pybiomart"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b29d125",
   "metadata": {},
   "source": [
    "# 0. user-defined variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e34ff5cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_threshold = 2\n",
    "discrete_fc_threshold = 1\n",
    "noise_threshold = 1/2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "262af060",
   "metadata": {},
   "outputs": [],
   "source": [
    "hypotheses = ['hypothesis_A', 'hypothesis_B', 'hypothesis_C', 'hypothesis_D', 'hypothesis_E']\n",
    "trends = ['up', 'down']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a70af446",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_file = '/home/adrian/projects/hegoi/metadata/hegoi metadata - hypotheses formatted for filter.tsv'\n",
    "tpm_file = '/home/adrian/projects/hegoi/results/tpm/DESeq2_TPM_values.tsv'\n",
    "deseq2_folder = '/home/adrian/projects/hegoi/results/subsamples/DESeq2/'\n",
    "\n",
    "output_dir = '/home/adrian/projects/hegoi/results/subsamples/DESeq2/filtered/'\n",
    "filtered_DEGs_file = '/home/adrian/projects/hegoi/results/subsamples/DESeq2/filtered/filtered_DEGs.tsv'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de5d0737",
   "metadata": {},
   "source": [
    "# 1. build annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "515f1a71",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NCBI gene (formerly Entrezgene) ID</th>\n",
       "      <th>Gene type</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene description</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000210049</th>\n",
       "      <td>-1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TF</td>\n",
       "      <td>mitochondrially encoded tRNA-Phe (UUU/C) [Sour...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000211459</th>\n",
       "      <td>-1</td>\n",
       "      <td>Mt_rRNA</td>\n",
       "      <td>MT-RNR1</td>\n",
       "      <td>mitochondrially encoded 12S rRNA [Source:HGNC ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000210077</th>\n",
       "      <td>-1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TV</td>\n",
       "      <td>mitochondrially encoded tRNA-Val (GUN) [Source...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000210082</th>\n",
       "      <td>-1</td>\n",
       "      <td>Mt_rRNA</td>\n",
       "      <td>MT-RNR2</td>\n",
       "      <td>mitochondrially encoded 16S rRNA [Source:HGNC ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000209082</th>\n",
       "      <td>-1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TL1</td>\n",
       "      <td>mitochondrially encoded tRNA-Leu (UUA/G) 1 [So...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000269732</th>\n",
       "      <td>-1</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>WBP1LP7</td>\n",
       "      <td>WBP1L pseudogene 7 [Source:HGNC Symbol;Acc:HGN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000284733</th>\n",
       "      <td>729759</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>olfactory receptor family 4 subfamily F member...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000233653</th>\n",
       "      <td>-1</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>CICP7</td>\n",
       "      <td>capicua transcriptional repressor pseudogene 7...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000250575</th>\n",
       "      <td>-1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>novel pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000117682</th>\n",
       "      <td>79947</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>DHDDS</td>\n",
       "      <td>dehydrodolichyl diphosphate synthase subunit [...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>68664 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 NCBI gene (formerly Entrezgene) ID               Gene type  \\\n",
       "Gene stable ID                                                                \n",
       "ENSG00000210049                                  -1                 Mt_tRNA   \n",
       "ENSG00000211459                                  -1                 Mt_rRNA   \n",
       "ENSG00000210077                                  -1                 Mt_tRNA   \n",
       "ENSG00000210082                                  -1                 Mt_rRNA   \n",
       "ENSG00000209082                                  -1                 Mt_tRNA   \n",
       "...                                             ...                     ...   \n",
       "ENSG00000269732                                  -1    processed_pseudogene   \n",
       "ENSG00000284733                              729759          protein_coding   \n",
       "ENSG00000233653                                  -1    processed_pseudogene   \n",
       "ENSG00000250575                                  -1  unprocessed_pseudogene   \n",
       "ENSG00000117682                               79947          protein_coding   \n",
       "\n",
       "                Gene name                                   Gene description  \n",
       "Gene stable ID                                                                \n",
       "ENSG00000210049     MT-TF  mitochondrially encoded tRNA-Phe (UUU/C) [Sour...  \n",
       "ENSG00000211459   MT-RNR1  mitochondrially encoded 12S rRNA [Source:HGNC ...  \n",
       "ENSG00000210077     MT-TV  mitochondrially encoded tRNA-Val (GUN) [Source...  \n",
       "ENSG00000210082   MT-RNR2  mitochondrially encoded 16S rRNA [Source:HGNC ...  \n",
       "ENSG00000209082    MT-TL1  mitochondrially encoded tRNA-Leu (UUA/G) 1 [So...  \n",
       "...                   ...                                                ...  \n",
       "ENSG00000269732   WBP1LP7  WBP1L pseudogene 7 [Source:HGNC Symbol;Acc:HGN...  \n",
       "ENSG00000284733    OR4F29  olfactory receptor family 4 subfamily F member...  \n",
       "ENSG00000233653     CICP7  capicua transcriptional repressor pseudogene 7...  \n",
       "ENSG00000250575       NaN                                   novel pseudogene  \n",
       "ENSG00000117682     DHDDS  dehydrodolichyl diphosphate synthase subunit [...  \n",
       "\n",
       "[68664 rows x 4 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = pybiomart.Dataset(name='hsapiens_gene_ensembl', host='http://www.ensembl.org')\n",
    "annotation = dataset.query(attributes=['ensembl_gene_id', 'entrezgene_id', 'gene_biotype', 'external_gene_name', 'description'])\n",
    "annotation.fillna(value={'NCBI gene (formerly Entrezgene) ID':-1}, inplace=True)\n",
    "annotation = annotation.astype({'NCBI gene (formerly Entrezgene) ID':'int'})\n",
    "annotation.set_index('Gene stable ID', inplace=True)\n",
    "annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "810f329b",
   "metadata": {},
   "source": [
    "# 1. read input files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6be307aa",
   "metadata": {},
   "source": [
    "## 1.1 read expression data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9f210abc",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(40173, 26)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression = pandas.read_csv(tpm_file, sep='\\t', index_col=0)\n",
    "sample_names = expression.columns.to_list()\n",
    "\n",
    "expression.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2508a907",
   "metadata": {},
   "source": [
    "## 1.2. read metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "dbaaf8ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient</th>\n",
       "      <th>sampleA</th>\n",
       "      <th>sampleB</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hypothesis</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>hypothesis_A</th>\n",
       "      <td>46</td>\n",
       "      <td>Stat46</td>\n",
       "      <td>Lami46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hypothesis_A</th>\n",
       "      <td>153</td>\n",
       "      <td>Stat153</td>\n",
       "      <td>Lam153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hypothesis_A</th>\n",
       "      <td>154</td>\n",
       "      <td>Stat154</td>\n",
       "      <td>Lami154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hypothesis_A</th>\n",
       "      <td>94</td>\n",
       "      <td>Stat94</td>\n",
       "      <td>Lami94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hypothesis_B</th>\n",
       "      <td>46</td>\n",
       "      <td>Lami46</td>\n",
       "      <td>Osci46</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              patient  sampleA  sampleB\n",
       "hypothesis                             \n",
       "hypothesis_A       46   Stat46   Lami46\n",
       "hypothesis_A      153  Stat153   Lam153\n",
       "hypothesis_A      154  Stat154  Lami154\n",
       "hypothesis_A       94   Stat94   Lami94\n",
       "hypothesis_B       46   Lami46   Osci46"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata = pandas.read_csv(metadata_file, sep='\\t', index_col=0)\n",
    "metadata.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a409f8c2",
   "metadata": {},
   "source": [
    "# 2. iterate over hypotheses and patients"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "38737756",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['NCBI gene (formerly Entrezgene) ID', 'Gene type', 'Gene name',\n",
       "       'Gene description'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "61b071a9",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "working with hypothesis_A up\n",
      "['hypothesis_A_154_up.tsv', 'hypothesis_A_46_up.tsv', 'hypothesis_A_94_up.tsv', 'hypothesis_A_153_up.tsv'] 4\n",
      "\t WARNING: removing due to noise 0.6356594003488406\n",
      "\t ENSG00000115461 IGFBP5 insulin like growth factor binding protein 5 [Source:HGNC Symbol;Acc:HGNC:5474]\n",
      "\t ref [0. 0. 0. 1.] 0\n",
      "\t sam [423.   7.   4. 105.] 0.6356594003488406\n",
      "\n",
      "\t WARNING: removing due to noise 0.5715217447490105\n",
      "\t ENSG00000187922 LCN10 lipocalin 10 [Source:HGNC Symbol;Acc:HGNC:20892]\n",
      "\t ref [0. 0. 0. 0.] 0\n",
      "\t sam [130.   1.   3.  49.] 0.5715217447490105\n",
      "\n",
      "\t WARNING: removing due to noise 0.5411718220815724\n",
      "\t ENSG00000198643 FAM3D FAM3 metabolism regulating signaling molecule D [Source:HGNC Symbol;Acc:HGNC:18665]\n",
      "\t ref [0. 0. 0.] 0\n",
      "\t sam [41.  1. 12.] 0.5411718220815724\n",
      "\n",
      "\t WARNING: removing due to noise 0.6214059902060192\n",
      "\t ENSG00000159167 STC1 stanniocalcin 1 [Source:HGNC Symbol;Acc:HGNC:11373]\n",
      "\t ref [1. 1. 1. 0.] 0\n",
      "\t sam [85.  5.  7. 11.] 0.6214059902060192\n",
      "\n",
      "\t WARNING: removing due to noise 0.6137636568827022\n",
      "\t ENSG00000197632 SERPINB2 serpin family B member 2 [Source:HGNC Symbol;Acc:HGNC:8584]\n",
      "\t ref [0. 1. 0.] 0\n",
      "\t sam [15. 88.  3.] 0.6137636568827022\n",
      "\n",
      "DEGs union across patients \t 2453\n",
      "working set after n - 1 consistency \t 198\n",
      "working set after low-expression filter \t 190\n",
      "working set after discrete FC filter \t 190\n",
      "working set after noise filter \t 185\n",
      "(185, 5)\n",
      "about to store a dataframe of shape (185, 9)\n",
      "--------------------------------------\n",
      "working with hypothesis_A down\n",
      "['hypothesis_A_154_down.tsv', 'hypothesis_A_46_down.tsv', 'hypothesis_A_94_down.tsv', 'hypothesis_A_153_down.tsv'] 4\n",
      "\t WARNING: removing due to noise 0.7956726178163475\n",
      "\t ENSG00000206503 HLA-A major histocompatibility complex, class I, A [Source:HGNC Symbol;Acc:HGNC:4931]\n",
      "\t ref [231.   3.   1.] 0.7956726178163475\n",
      "\t sam [0. 1. 0.] 0\n",
      "\n",
      "\t WARNING: removing due to noise 0.697614984548545\n",
      "\t ENSG00000178882 Gene stable ID\n",
      "ENSG00000178882    RFLNA\n",
      "ENSG00000178882    RFLNA\n",
      "Name: Gene name, dtype: object Gene stable ID\n",
      "ENSG00000178882    refilin A [Source:HGNC Symbol;Acc:HGNC:27051]\n",
      "ENSG00000178882    refilin A [Source:HGNC Symbol;Acc:HGNC:27051]\n",
      "Name: Gene description, dtype: object\n",
      "\t ref [18. 26.  5.] 0.3058956075634982\n",
      "\t sam [0. 9. 1.] 0.697614984548545\n",
      "\n",
      "\t WARNING: removing due to noise 0.6571287406727709\n",
      "\t ENSG00000170323 FABP4 fatty acid binding protein 4 [Source:HGNC Symbol;Acc:HGNC:3559]\n",
      "\t ref [ 45.  62. 101.  38.] 0.19852220239145982\n",
      "\t sam [ 1. 18.  1.  2.] 0.6571287406727709\n",
      "\n",
      "\t WARNING: removing due to noise 0.5034040503016264\n",
      "\t ENSG00000163739 CXCL1 C-X-C motif chemokine ligand 1 [Source:HGNC Symbol;Acc:HGNC:4602]\n",
      "\t ref [369. 966. 458.] 0.2540339620380326\n",
      "\t sam [ 11. 197.  62.] 0.5034040503016264\n",
      "\n",
      "\t WARNING: removing due to noise 0.6093413965251636\n",
      "\t ENSG00000133110 POSTN periostin [Source:HGNC Symbol;Acc:HGNC:16953]\n",
      "\t ref [1198. 3440.  881. 2108.] 0.2604452016784179\n",
      "\t sam [ 46.  49.   8. 353.] 0.6093413965251636\n",
      "\n",
      "\t WARNING: removing due to noise 0.6319062870042959\n",
      "\t ENSG00000187513 GJA4 gap junction protein alpha 4 [Source:HGNC Symbol;Acc:HGNC:4278]\n",
      "\t ref [ 67. 179. 241. 110.] 0.22227464877679073\n",
      "\t sam [ 3. 19.  0.  2.] 0.6319062870042959\n",
      "\n",
      "\t WARNING: removing due to noise 0.5351001587841251\n",
      "\t ENSG00000134363 FST follistatin [Source:HGNC Symbol;Acc:HGNC:3971]\n",
      "\t ref [16. 44.  9. 18.] 0.3051394012059376\n",
      "\t sam [ 1. 12.  3.  1.] 0.5351001587841251\n",
      "\n",
      "\t WARNING: removing due to noise 0.5759050847723663\n",
      "\t ENSG00000139329 LUM lumican [Source:HGNC Symbol;Acc:HGNC:6724]\n",
      "\t ref [ 3.  1. 16.] 0.5759050847723663\n",
      "\t sam [0. 0. 5.] 0\n",
      "\n",
      "\t WARNING: removing due to noise 0.6295798667988843\n",
      "\t ENSG00000137959 IFI44L interferon induced protein 44 like [Source:HGNC Symbol;Acc:HGNC:17817]\n",
      "\t ref [ 7.  3. 55.] 0.6295798667988843\n",
      "\t sam [1. 0. 3.] 0\n",
      "\n",
      "\t WARNING: removing due to noise 0.6692657668420422\n",
      "\t ENSG00000265107 GJA5 gap junction protein alpha 5 [Source:HGNC Symbol;Acc:HGNC:4279]\n",
      "\t ref [  3. 125.  18.] 0.6444425391579529\n",
      "\t sam [1. 7. 0.] 0.6692657668420422\n",
      "\n",
      "\t WARNING: removing due to noise 0.5229763603684907\n",
      "\t ENSG00000168542 COL3A1 collagen type III alpha 1 chain [Source:HGNC Symbol;Acc:HGNC:2201]\n",
      "\t ref [ 29.  22. 113.] 0.4366750600510555\n",
      "\t sam [ 9.  1. 29.] 0.5229763603684907\n",
      "\n",
      "\t WARNING: removing due to noise 0.6152009608430583\n",
      "\t ENSG00000285438 SOX7 SRY-box transcription factor 7 [Source:HGNC Symbol;Acc:HGNC:18196]\n",
      "\t ref [ 6. 21. 51.] 0.415432096051786\n",
      "\t sam [ 1.  3. 20.] 0.6152009608430583\n",
      "\n",
      "DEGs union across patients \t 2984\n",
      "working set after n - 1 consistency \t 373\n",
      "working set after low-expression filter \t 329\n",
      "working set after discrete FC filter \t 329\n",
      "working set after noise filter \t 317\n",
      "(317, 5)\n",
      "about to store a dataframe of shape (318, 9)\n",
      "--------------------------------------\n",
      "working with hypothesis_B up\n",
      "['hypothesis_B_154_up.tsv', 'hypothesis_B_46_up.tsv', 'hypothesis_B_153_up.tsv'] 3\n",
      "\t WARNING: removing due to noise 0.6280742930213279\n",
      "\t ENSG00000091986 CCDC80 coiled-coil domain containing 80 [Source:HGNC Symbol;Acc:HGNC:30649]\n",
      "\t ref [ 1.  1. 11.] 0.6280742930213279\n",
      "\t sam [ 9.  7. 48.] 0.5107885549977278\n",
      "\n",
      "\t WARNING: removing due to noise 0.6060915267313264\n",
      "\t ENSG00000169429 CXCL8 C-X-C motif chemokine ligand 8 [Source:HGNC Symbol;Acc:HGNC:6025]\n",
      "\t ref [  8. 104.] 0.6060915267313264\n",
      "\t sam [ 49. 216.] 0.4456106885213337\n",
      "\n",
      "\t WARNING: removing due to noise 0.5568051909067913\n",
      "\t ENSG00000133110 POSTN periostin [Source:HGNC Symbol;Acc:HGNC:16953]\n",
      "\t ref [ 46.  49. 353.] 0.5568051909067913\n",
      "\t sam [ 267.  281. 1190.] 0.4303661571281814\n",
      "\n",
      "\t WARNING: removing due to noise 0.5303300858899106\n",
      "\t ENSG00000150594 ADRA2A adrenoceptor alpha 2A [Source:HGNC Symbol;Acc:HGNC:281]\n",
      "\t ref [2. 0.] 0\n",
      "\t sam [7. 1.] 0.5303300858899106\n",
      "\n",
      "\t WARNING: removing due to noise 0.5303300858899106\n",
      "\t ENSG00000172260 NEGR1 neuronal growth regulator 1 [Source:HGNC Symbol;Acc:HGNC:17302]\n",
      "\t ref [0. 0.] 0\n",
      "\t sam [7. 1.] 0.5303300858899106\n",
      "\n",
      "\t WARNING: removing due to noise 0.7071067811865475\n",
      "\t ENSG00000140092 FBLN5 fibulin 5 [Source:HGNC Symbol;Acc:HGNC:3602]\n",
      "\t ref [0. 5.] 0.7071067811865475\n",
      "\t sam [ 2. 16.] 0.5499719409228703\n",
      "\n",
      "\t WARNING: removing due to noise 0.5050762722761053\n",
      "\t ENSG00000128606 LRRC17 leucine rich repeat containing 17 [Source:HGNC Symbol;Acc:HGNC:16895]\n",
      "\t ref [0. 1.] 0\n",
      "\t sam [1. 6.] 0.5050762722761053\n",
      "\n",
      "\t WARNING: removing due to noise 0.656599153958937\n",
      "\t ENSG00000277236 CCL14 C-C motif chemokine ligand 14 [Source:HGNC Symbol;Acc:HGNC:10612]\n",
      "\t ref [0. 0.] 0\n",
      "\t sam [ 2. 54.] 0.656599153958937\n",
      "\n",
      "DEGs union across patients \t 1335\n",
      "working set after n - 1 consistency \t 250\n",
      "working set after low-expression filter \t 218\n",
      "working set after discrete FC filter \t 217\n",
      "working set after noise filter \t 209\n",
      "(209, 5)\n",
      "about to store a dataframe of shape (209, 9)\n",
      "--------------------------------------\n",
      "working with hypothesis_B down\n",
      "['hypothesis_B_46_down.tsv', 'hypothesis_B_154_down.tsv', 'hypothesis_B_153_down.tsv'] 3\n",
      "\t WARNING: removing due to noise 0.5411718220815724\n",
      "\t ENSG00000198643 FAM3D FAM3 metabolism regulating signaling molecule D [Source:HGNC Symbol;Acc:HGNC:18665]\n",
      "\t ref [ 1. 41. 12.] 0.5411718220815724\n",
      "\t sam [0. 1. 0.] 0\n",
      "\n",
      "\t WARNING: removing due to noise 0.5087602374750728\n",
      "\t ENSG00000267206 LCN6 lipocalin 6 [Source:HGNC Symbol;Acc:HGNC:17337]\n",
      "\t ref [ 21. 367. 111.] 0.5087602374750728\n",
      "\t sam [1. 1. 3.] 0\n",
      "\n",
      "\t WARNING: removing due to noise 0.5195887333176439\n",
      "\t ENSG00000095713 CRTAC1 cartilage acidic protein 1 [Source:HGNC Symbol;Acc:HGNC:14882]\n",
      "\t ref [ 3. 40. 22.] 0.402555936549546\n",
      "\t sam [0. 8. 3.] 0.5195887333176439\n",
      "\n",
      "\t WARNING: removing due to noise 0.5122571692103561\n",
      "\t ENSG00000187922 LCN10 lipocalin 10 [Source:HGNC Symbol;Acc:HGNC:20892]\n",
      "\t ref [  1. 130.  49.] 0.5122571692103561\n",
      "\t sam [0. 0. 0.] 0\n",
      "\n",
      "\t WARNING: removing due to noise 0.6331072343181879\n",
      "\t ENSG00000196154 S100A4 S100 calcium binding protein A4 [Source:HGNC Symbol;Acc:HGNC:10494]\n",
      "\t ref [  9. 163.] 0.6331072343181879\n",
      "\t sam [2. 7.] 0.39283710065919303\n",
      "\n",
      "\t WARNING: removing due to noise 0.5748831023947797\n",
      "\t ENSG00000115461 IGFBP5 insulin like growth factor binding protein 5 [Source:HGNC Symbol;Acc:HGNC:5474]\n",
      "\t ref [  7. 423. 105.] 0.5748831023947797\n",
      "\t sam [2. 4. 3.] 0.15713484026367724\n",
      "\n",
      "\t WARNING: removing due to noise 0.5303300858899106\n",
      "\t ENSG00000138061 CYP1B1 cytochrome P450 family 1 subfamily B member 1 [Source:HGNC Symbol;Acc:HGNC:2597]\n",
      "\t ref [39. 65.] 0.17677669529663687\n",
      "\t sam [1. 7.] 0.5303300858899106\n",
      "\n",
      "\t WARNING: removing due to noise 0.5450614771646304\n",
      "\t ENSG00000159167 STC1 stanniocalcin 1 [Source:HGNC Symbol;Acc:HGNC:11373]\n",
      "\t ref [85. 11.] 0.5450614771646304\n",
      "\t sam [2. 1.] 0\n",
      "\n",
      "\t WARNING: removing due to noise 0.565685424949238\n",
      "\t ENSG00000132470 ITGB4 integrin subunit beta 4 [Source:HGNC Symbol;Acc:HGNC:6158]\n",
      "\t ref [151. 102.] 0.1369495346962088\n",
      "\t sam [9. 1.] 0.565685424949238\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEGs union across patients \t 1414\n",
      "working set after n - 1 consistency \t 356\n",
      "working set after low-expression filter \t 340\n",
      "working set after discrete FC filter \t 339\n",
      "working set after noise filter \t 330\n",
      "(330, 5)\n",
      "about to store a dataframe of shape (330, 9)\n",
      "--------------------------------------\n",
      "working with hypothesis_C up\n",
      "['hypothesis_C_46_up.tsv', 'hypothesis_C_154_up.tsv', 'hypothesis_C_153_up.tsv', 'hypothesis_C_109_up.tsv'] 4\n",
      "\t WARNING: removing due to noise 0.5665577237325317\n",
      "\t ENSG00000115457 IGFBP2 insulin like growth factor binding protein 2 [Source:HGNC Symbol;Acc:HGNC:5471]\n",
      "\t ref [0. 2. 7.] 0.5665577237325317\n",
      "\t sam [ 0. 13. 34.] 0.516222422271712\n",
      "\n",
      "\t WARNING: removing due to noise 0.5400617248673217\n",
      "\t ENSG00000273622 CDC42EP5 CDC42 effector protein 5 [Source:HGNC Symbol;Acc:HGNC:17408]\n",
      "\t ref [6. 2. 0.] 0.5400617248673217\n",
      "\t sam [75. 38. 34.] 0.21747438589313992\n",
      "\n",
      "\t WARNING: removing due to noise 0.5084322977157767\n",
      "\t ENSG00000100321 SYNGR1 synaptogyrin 1 [Source:HGNC Symbol;Acc:HGNC:11498]\n",
      "\t ref [5. 0. 2.] 0.5084322977157767\n",
      "\t sam [15.  3.  9.] 0.31426968052735443\n",
      "\n",
      "\t WARNING: removing due to noise 0.664873891551373\n",
      "\t ENSG00000120708 TGFBI transforming growth factor beta induced [Source:HGNC Symbol;Acc:HGNC:11771]\n",
      "\t ref [28.  1.  3.] 0.664873891551373\n",
      "\t sam [61. 10. 24.] 0.3922612429658947\n",
      "\n",
      "\t WARNING: removing due to noise 0.7412686197773832\n",
      "\t ENSG00000276409 CCL14 C-C motif chemokine ligand 14 [Source:HGNC Symbol;Acc:HGNC:10612]\n",
      "\t ref [ 0. 15.  1.] 0.7412686197773832\n",
      "\t sam [217. 116. 130.] 0.16714141683225217\n",
      "\n",
      "DEGs union across patients \t 5240\n",
      "working set after n - 1 consistency \t 193\n",
      "working set after low-expression filter \t 103\n",
      "working set after discrete FC filter \t 101\n",
      "working set after noise filter \t 96\n",
      "(96, 5)\n",
      "about to store a dataframe of shape (97, 9)\n",
      "--------------------------------------\n",
      "working with hypothesis_C down\n",
      "['hypothesis_C_46_down.tsv', 'hypothesis_C_154_down.tsv', 'hypothesis_C_153_down.tsv', 'hypothesis_C_109_down.tsv'] 4\n",
      "\t WARNING: removing due to noise 0.5794254758208154\n",
      "\t ENSG00000187922 LCN10 lipocalin 10 [Source:HGNC Symbol;Acc:HGNC:20892]\n",
      "\t ref [ 1.  5. 20.  1.] 0.5794254758208154\n",
      "\t sam [0. 2. 0. 0.] 0\n",
      "\n",
      "\t WARNING: removing due to noise 0.6075738683564674\n",
      "\t ENSG00000267206 LCN6 lipocalin 6 [Source:HGNC Symbol;Acc:HGNC:17337]\n",
      "\t ref [ 34.  47. 127.] 0.34241661332486173\n",
      "\t sam [0. 9. 2.] 0.6075738683564674\n",
      "\n",
      "\t WARNING: removing due to noise 0.5212165257037297\n",
      "\t ENSG00000111012 CYP27B1 cytochrome P450 family 27 subfamily B member 1 [Source:HGNC Symbol;Acc:HGNC:2606]\n",
      "\t ref [ 2. 30.  8.] 0.5212165257037297\n",
      "\t sam [0. 0. 0.] 0\n",
      "\n",
      "\t WARNING: removing due to noise 0.5366269111911884\n",
      "\t ENSG00000123496 IL13RA2 interleukin 13 receptor subunit alpha 2 [Source:HGNC Symbol;Acc:HGNC:5975]\n",
      "\t ref [ 46. 111. 130.] 0.21705573910766432\n",
      "\t sam [10.  1.  2.] 0.5366269111911884\n",
      "\n",
      "\t WARNING: removing due to noise 0.7041346661211583\n",
      "\t ENSG00000115461 IGFBP5 insulin like growth factor binding protein 5 [Source:HGNC Symbol;Acc:HGNC:5474]\n",
      "\t ref [ 5. 99.  5.] 0.7041346661211583\n",
      "\t sam [2. 1. 1.] 0\n",
      "\n",
      "DEGs union across patients \t 4689\n",
      "working set after n - 1 consistency \t 75\n",
      "working set after low-expression filter \t 55\n",
      "working set after discrete FC filter \t 54\n",
      "working set after noise filter \t 49\n",
      "(49, 5)\n",
      "about to store a dataframe of shape (49, 9)\n",
      "--------------------------------------\n",
      "working with hypothesis_D up\n",
      "['hypothesis_D_153_up.tsv', 'hypothesis_D_94_up.tsv', 'hypothesis_D_46_up.tsv', 'hypothesis_D_154_up.tsv'] 4\n",
      "DEGs union across patients \t 3681\n",
      "working set after n - 1 consistency \t 16\n",
      "working set after low-expression filter \t 5\n",
      "working set after discrete FC filter \t 5\n",
      "working set after noise filter \t 5\n",
      "(5, 5)\n",
      "about to store a dataframe of shape (5, 9)\n",
      "--------------------------------------\n",
      "working with hypothesis_D down\n",
      "['hypothesis_D_94_down.tsv', 'hypothesis_D_154_down.tsv', 'hypothesis_D_153_down.tsv', 'hypothesis_D_46_down.tsv'] 4\n",
      "\t WARNING: removing due to noise 0.8164965809277261\n",
      "\t ENSG00000164530 PI16 peptidase inhibitor 16 [Source:HGNC Symbol;Acc:HGNC:21245]\n",
      "\t ref [113.  60.   0.] 0.46216273716646816\n",
      "\t sam [ 0. 12.  0.] 0.8164965809277261\n",
      "\n",
      "DEGs union across patients \t 4549\n",
      "working set after n - 1 consistency \t 48\n",
      "working set after low-expression filter \t 22\n",
      "working set after discrete FC filter \t 22\n",
      "working set after noise filter \t 21\n",
      "(21, 5)\n",
      "about to store a dataframe of shape (21, 9)\n",
      "--------------------------------------\n",
      "working with hypothesis_E up\n",
      "['hypothesis_E_154_up.tsv', 'hypothesis_E_109_up.tsv', 'hypothesis_E_46_up.tsv', 'hypothesis_E_153_up.tsv'] 4\n",
      "DEGs union across patients \t 1257\n",
      "working set after n - 1 consistency \t 2\n",
      "working set after low-expression filter \t 2\n",
      "working set after discrete FC filter \t 2\n",
      "working set after noise filter \t 2\n",
      "(2, 5)\n",
      "about to store a dataframe of shape (2, 9)\n",
      "--------------------------------------\n",
      "working with hypothesis_E down\n",
      "['hypothesis_E_46_down.tsv', 'hypothesis_E_109_down.tsv', 'hypothesis_E_153_down.tsv', 'hypothesis_E_154_down.tsv'] 4\n",
      "DEGs union across patients \t 1509\n",
      "working set after n - 1 consistency \t 13\n",
      "working set after low-expression filter \t 9\n",
      "working set after discrete FC filter \t 9\n",
      "working set after noise filter \t 9\n",
      "(9, 5)\n",
      "about to store a dataframe of shape (9, 9)\n",
      "--------------------------------------\n"
     ]
    }
   ],
   "source": [
    "### 1. iterate over hypothesis\n",
    "for hypothesis in hypotheses:\n",
    "    \n",
    "    ### 2. iterate over trends\n",
    "    for trend in trends:\n",
    "        \n",
    "        print('working with {} {}'.format(hypothesis, trend))\n",
    "        box = {}; predf = {}\n",
    "        \n",
    "        ### 3. gather patients\n",
    "        elements = os.listdir(deseq2_folder)\n",
    "        working_elements = [element for element in elements if hypothesis in element and trend in element]\n",
    "        number_of_patients = len(working_elements)\n",
    "        \n",
    "        print(working_elements, len(working_elements))\n",
    "\n",
    "        ### 4. gather information\n",
    "        \n",
    "        ## 4.1. get adjusted P values and log2FC\n",
    "        for working_element in working_elements:\n",
    "            \n",
    "            ## 4.1. define reference and sample expression labels\n",
    "            patient = int(working_element.split('_')[2])\n",
    "            ref_label = None; sam_label = None\n",
    "            ref_label = metadata[metadata['patient'] == patient].loc[hypothesis]['sampleA']\n",
    "            sam_label = metadata[metadata['patient'] == patient].loc[hypothesis]['sampleB']\n",
    "            \n",
    "            ## 4.2. get adjusted P values and log2FC\n",
    "            f = open(deseq2_folder + working_element, 'r')\n",
    "            next(f)\n",
    "            for line in f:\n",
    "                v = line.split('\\t')\n",
    "                ensembl = v[0]\n",
    "                log2FC = float(v[2])\n",
    "                adjusted = float(v[6])\n",
    "                                \n",
    "                if ensembl not in box:\n",
    "                    box[ensembl]=[[log2FC], [adjusted], [], [], []]\n",
    "                else:\n",
    "                    box[ensembl][0].append(log2FC); box[ensembl][1].append(adjusted)   \n",
    "\n",
    "                ### 4.3. get reference and sample expression   \n",
    "                tpm_ref = expression[ref_label].loc[ensembl]\n",
    "                tpm_sam = expression[sam_label].loc[ensembl]\n",
    "\n",
    "                num = numpy.around(tpm_sam) + 1\n",
    "                den = numpy.around(tpm_ref) + 1\n",
    "                fc = num/den\n",
    "                abs_discrete_log2FC = numpy.abs(numpy.log2(fc))\n",
    "\n",
    "                box[ensembl][2].append(tpm_ref)\n",
    "                box[ensembl][3].append(tpm_sam)\n",
    "                box[ensembl][4].append(abs_discrete_log2FC)\n",
    "                \n",
    "            # late file closure\n",
    "            f.close()\n",
    "        \n",
    "        ### 5. perform filters\n",
    "        calls = len(box.keys())\n",
    "        consistency_filter_count = 0\n",
    "        low_expression_filter_count = 0\n",
    "        discrete_filter_count = 0\n",
    "        noise_filter_count = 0\n",
    "        \n",
    "        for gene in box:\n",
    "            \n",
    "            selected = True\n",
    "                        \n",
    "            ## F1 | consistency across patients, DEG called at least n - 1\n",
    "            if len(box[gene][0]) < number_of_patients - 1:\n",
    "                selected = False\n",
    "                consistency_filter_count = consistency_filter_count + 1\n",
    "                \n",
    "            ## F2 | filter on low expression\n",
    "            if selected == True:\n",
    "                a = numpy.median(box[gene][2])\n",
    "                b = numpy.median(box[gene][3])\n",
    "                top = numpy.max([a, b])\n",
    "                if top < expression_threshold:\n",
    "                    selected = False\n",
    "                    low_expression_filter_count = low_expression_filter_count + 1\n",
    "            \n",
    "            ## F3 | filter on abs discrete log2 FC\n",
    "            if selected == True:\n",
    "                if numpy.median(box[gene][4]) < discrete_fc_threshold:\n",
    "                    selected = False\n",
    "                    discrete_filter_count = discrete_filter_count + 1\n",
    "                    \n",
    "            ## F4 | filter on noise \n",
    "            if selected == True:\n",
    "                \n",
    "                a = box[gene][2]; b = box[gene][3]\n",
    "                a = numpy.around(a); b = numpy.around(b)\n",
    "                \n",
    "                if numpy.mean(a) > 2:\n",
    "                    sem_ref = numpy.std(a) / numpy.sqrt(len(a))\n",
    "                    rsem_ref = sem_ref / numpy.mean(a)\n",
    "                else:\n",
    "                    rsem_ref = 0\n",
    "                \n",
    "                if numpy.mean(b) > 2:\n",
    "                    sem_sam = numpy.std(b) / numpy.sqrt(len(b))\n",
    "                    rsem_sam = sem_sam / numpy.mean(b)\n",
    "                else:\n",
    "                    rsem_sam = 0\n",
    "                \n",
    "                noise = numpy.max([rsem_ref, rsem_sam])\n",
    "                \n",
    "                if noise > noise_threshold:\n",
    "                    selected = False\n",
    "                    noise_filter_count = noise_filter_count + 1 \n",
    "                    print('\\t', 'WARNING: removing due to noise {}'.format(noise))\n",
    "                    print('\\t', gene, annotation.loc[gene, 'Gene name'], annotation.loc[gene, 'Gene description'])\n",
    "                    print('\\t', 'ref', a, rsem_ref)\n",
    "                    print('\\t', 'sam', b, rsem_sam)\n",
    "                    print()\n",
    "            ###\n",
    "            if selected == True:\n",
    "                predf[gene] = []\n",
    "                predf[gene].append(numpy.median(box[gene][0])) # log2FC\n",
    "                predf[gene].append(numpy.median(box[gene][1])) # adjusted P\n",
    "                predf[gene].append(numpy.median(box[gene][2])) # ref\n",
    "                predf[gene].append(numpy.median(box[gene][3])) # sam\n",
    "                predf[gene].append(numpy.median(box[gene][4])) # abs discrete log2FC                  \n",
    "                    \n",
    "        ### final print        \n",
    "        print('DEGs union across patients \\t {}'.format(calls))\n",
    "        print('working set after n - 1 consistency \\t {}'.format(calls-consistency_filter_count))\n",
    "        print('working set after low-expression filter \\t {}'.format(calls-consistency_filter_count-low_expression_filter_count))\n",
    "        print('working set after discrete FC filter \\t {}'.format(calls-consistency_filter_count-low_expression_filter_count-discrete_filter_count))\n",
    "        print('working set after noise filter \\t {}'.format(calls-consistency_filter_count-low_expression_filter_count-discrete_filter_count-noise_filter_count))\n",
    "        \n",
    "        ### generate a dataframe and store\n",
    "        outputfile = output_dir + 'filtered_' + hypothesis + '_' + trend + '.tsv'\n",
    "        \n",
    "        df = pandas.DataFrame.from_dict(predf)\n",
    "        dft = df.transpose()\n",
    "        dft.index.name = 'ENSEMBL'\n",
    "        dft.columns = ['log2FC', 'adjusted P', 'Reference expression (TPM)', 'Sample expression (TPM)', 'Discrete abs(log2FC)']\n",
    "        dft.sort_values(by='Discrete abs(log2FC)', axis=0, inplace=True, ascending=False)\n",
    "        print(dft.shape)\n",
    "        \n",
    "        fdf = pandas.merge(dft, annotation.loc[dft.index, ], left_index=True, right_index=True)\n",
    "        fdf['Gene description'] = fdf['Gene description'].str.split(' \\[Source').str.get(0)\n",
    "        fdf = fdf[['NCBI gene (formerly Entrezgene) ID', 'Gene type', 'Gene name', 'Gene description', 'log2FC', 'adjusted P', 'Reference expression (TPM)', 'Sample expression (TPM)', 'Discrete abs(log2FC)']]\n",
    "        print('about to store a dataframe of shape {}'.format(fdf.shape))\n",
    "        \n",
    "        fdf.to_csv(outputfile)\n",
    "        \n",
    "        print('--------------------------------------')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
